File: services/nuldoc/nuldoc-src/components/PageLayout.ts
import { Config } from "../config.ts";
import { elem, Element, Node } from "../dom.ts";
import StaticStylesheet from "./StaticStylesheet.ts";

type Props = {
  metaCopyrightYear: number;
  metaDescription: string;
  metaKeywords?: string[];
  metaTitle: string;
  metaAtomFeedHref?: string;
  requiresSyntaxHighlight?: boolean;
  site: "default" | "about" | "blog" | "slides";
  config: Config;
  children: Node;
};

export default async function PageLayout(
  {
    metaCopyrightYear,
    metaDescription,
    metaKeywords,
    metaTitle,
    metaAtomFeedHref,
    requiresSyntaxHighlight: _,
    site,
    config,
    children,
  }: Props,
): Promise<Element> {
  return elem(
    "html",
    { lang: "ja-JP" },
    elem(
      "head",
      {},
      elem("meta", { charset: "UTF-8" }),
      elem("meta", {
        name: "viewport",
        content: "width=device-width, initial-scale=1.0",
      }),
      elem("meta", { name: "author", content: config.site.author }),
      elem("meta", {
        name: "copyright",
        content: `&copy; ${metaCopyrightYear} ${config.site.author}`,
      }),
      elem("meta", { name: "description", content: metaDescription }),
      metaKeywords && metaKeywords.length !== 0
        ? elem("meta", { name: "keywords", content: metaKeywords.join(",") })
        : null,
      elem("meta", { property: "og:type", content: "article" }),
      elem("meta", { property: "og:title", content: metaTitle }),
      elem("meta", { property: "og:description", content: metaDescription }),
      elem("meta", {
        property: "og:site_name",
        content: config.sites[site].siteName,
      }),
      elem("meta", { property: "og:locale", content: "ja_JP" }),
      elem("meta", { name: "Hatena::Bookmark", content: "nocomment" }),
      metaAtomFeedHref
        ? elem("link", {
          rel: "alternate",
          href: metaAtomFeedHref,
          type: "application/atom+xml",
        })
        : null,
      elem("link", {
        rel: "icon",
        href: "/favicon.svg",
        type: "image/svg+xml",
      }),
      elem("title", {}, metaTitle),
      await StaticStylesheet({ fileName: "/style.css", config }),
    ),
    children,
  );
}

File: services/nuldoc/lib/nuldoc/components/page_layout.rb
module Nuldoc
  module Components
    class PageLayout < DOM::HTMLBuilder
      def initialize(meta_copyright_year:, meta_description:, meta_title:, site:, config:, children:,
                     meta_keywords: nil, meta_atom_feed_href: nil)
        super()
        @meta_copyright_year = meta_copyright_year
        @meta_description = meta_description
        @meta_title = meta_title
        @site = site
        @config = config
        @children = children
        @meta_keywords = meta_keywords
        @meta_atom_feed_href = meta_atom_feed_href
      end

      def build
        site_entry = @config.site_entry(@site)

        html(lang: 'ja-JP') do
          head do
            meta(charset: 'UTF-8')
            meta(name: 'viewport', content: 'width=device-width, initial-scale=1.0')
            meta(name: 'author', content: @config.site.author)
            meta(name: 'copyright', content: "&copy; #{@meta_copyright_year} #{@config.site.author}")
            meta(name: 'description', content: @meta_description)
            if @meta_keywords && !@meta_keywords.empty?
              meta(name: 'keywords',
                   content: @meta_keywords.join(','))
            end
            meta(property: 'og:type', content: 'article')
            meta(property: 'og:title', content: @meta_title)
            meta(property: 'og:description', content: @meta_description)
            meta(property: 'og:site_name', content: site_entry.site_name)
            meta(property: 'og:locale', content: 'ja_JP')
            meta(name: 'Hatena::Bookmark', content: 'nocomment')
            link(rel: 'alternate', href: @meta_atom_feed_href, type: 'application/atom+xml') if @meta_atom_feed_href
            link(rel: 'icon', href: '/favicon.svg', type: 'image/svg+xml')
            title { text @meta_title }
            render(StaticStylesheet, file_name: '/style.css', config: @config)
          end
          child @children
        end
      end
    end
  end
end
